asdf/compression.py

```diff
@@ -30,118 +34,232 @@ def validate(compression):
+            block = memoryview(block).cast('c')  # don't copy on slice
+
+            while len(block):
+                if not self._size:
+                    # Don't know the (compressed) length of this block yet
+                    if len(self._partial_len) + len(block) < 4:
+                        self._partial_len += block
+                        break  # we've exhausted the block
+                    if self._partial_len:
+                        # If we started to fill a len key, finish filling it
+                        remaining = 4-len(self._partial_len)
+                        if remaining:
+                            self._partial_len += block[:remaining]
+                            block = block[remaining:]
+                        self._size = struct.unpack('!I', self._partial_len)[0]
+                        self._partial_len = b''
+                    else:
+                        # Otherwise just read the len key directly
+                        self._size = struct.unpack('!I', block[:4])[0]
+                        block = block[4:]
+
+                if len(block) < self._size or self._buffer is not None:
+                    # If we have a partial block, or we're already filling a buffer, use the buffer
+                    if self._buffer is None:
+                        self._buffer = np.empty(self._size, dtype=np.byte)  # use numpy instead of bytearray so we can avoid zero initialization
+                        self._pos = 0
+                    newbytes = min(self._size - self._pos, len(block))  # don't fill past the buffer len!
+                    self._buffer[self._pos:self._pos+newbytes] = np.frombuffer(block[:newbytes], dtype=np.byte)
+                    self._pos += newbytes
+                    block = block[newbytes:]
+
+                    if self._pos == self._size:
+                        _out = self._api.decompress(self._buffer, return_bytearray=True)
+                        out[bytesout:bytesout+len(_out)] = _out
+                        bytesout += len(_out)
+                        self._buffer = None
+                        self._size = 0
+                else:
+                    # We have at least one full block
+                    _out = self._api.decompress(memoryview(block[:self._size]), return_bytearray=True)
+                    out[bytesout:bytesout+len(_out)] = _out
+                    bytesout += len(_out)
+                    block = block[self._size:]
+                    self._size = 0
+
+        return bytesout
+
+    
+class ZlibCompressor:
+    def __init__(self, compression_kwargs=None, decompression_kwargs=None):
+        if compression_kwargs is None:
+            compression_kwargs = {}
+        if decompression_kwargs is None:
+            decompression_kwargs = {}
+            
+        self.compression_kwargs = compression_kwargs.copy()
+        self.decompression_kwargs = decompression_kwargs.copy()
+        
+    def compress(self, data, out=None):
```

Comment:
Did we drop this `out` argument from compress?

Reply:
(or is the ABC and LZ4Compressor missing an `out`)

Reply:
We could go either way, but let's drop `out` for now.